src/allmydata/test/web/test_web.py

```diff
@@ -657,55 +645,65 @@ def should302(self, res, which):
                       (which, res))
 
 
-
-class MultiFormatPageTests(TrialTestCase):
+class MultiFormatResourceTests(TrialTestCase):
     """
-    Tests for ``MultiFormatPage``.
+    Tests for ``MultiFormatResource``.
     """
+
     def resource(self):
         """
-        Create and return an instance of a ``MultiFormatPage`` subclass with two
-        formats: ``a`` and ``b``.
+        Create and return an instance of a ``MultiFormatResource`` subclass
+        with a default HTML format, and two custom formats: ``a`` and ``b``.
         """
-        class Content(MultiFormatPage):
-            docFactory = stan("doc factory")
+        class Content(MultiFormatResource):
+
+            def render_HTML(self, req):
+                return "html"
 
             def render_A(self, req):
                 return "a"
 
             def render_B(self, req):
                 return "b"
+
         return Content()
 
 
     def render(self, resource, **query_args):
         """
-        Render a Nevow ``Page`` against a request with the given query arguments.
+        Render a ``Resource`` against a request with the given query arguments.
 
         :param resource: The Nevow resource to render.
 
         :param query_args: The query arguments to put into the request being
             rendered.  A mapping from ``bytes`` to ``list`` of ``bytes``.
 
-        :return: The rendered response body as ``bytes``.
+        :return: The rendered response body as ``str``.
         """
-        ctx = WebContext(tag=resource)
-        req = FakeRequest(args=query_args)
-        ctx.remember(DefaultExceptionHandler(), ICanHandleException)
-        ctx.remember(req, IRequest)
-        ctx.remember(None, IData)
 
-        d = maybeDeferred(resource.renderHTTP, ctx)
-        d.addErrback(processingFailed, req, ctx)
-        res = self.successResultOf(d)
-        if isinstance(res, bytes):
-            return req.v + res
-        return req.v
+        # TODO: probably should: (1) refactor this out of here to a
+        # common module (test.common_web maybe?), and (2) replace
+        # nevow.inevow.IRequest with twisted.web.iweb.IRequest.  For
+        # (2) to happen, we will have to update web.common.get_arg()
+        # etc first.
+        from zope.interface import implementer
+        from nevow.inevow import IRequest
+        from twisted.web.server import Request
+        from twisted.web.test.requesthelper import DummyChannel
+
+        @implementer(IRequest)
+        class FakeRequest(Request):
+            def __init__(self, args):
+                Request.__init__(self, DummyChannel())
+                self.args = args
+                self.fields = dict()
+
+        return resource.render(FakeRequest(args=query_args))
```

Comment:
Is it worth changing this method at all?  Why not leave it as it was until we're ready to completely rip out Nevow?  The purpose here seems to be "render it like Nevow would  render it".  Using the Nevow APIs to do that seems like the most reliable solution.

Reply:
At this point all nevow `Page` and `MultiFormatPage` instances in Tahoe-LAFS have been removed/updated with Twisted `Resource` and `MultiFormatResource` instances.  Changes in this file re-purposes what formerly was `MultiFormatPageTests` as `MultiFormatResourceTests`, because `MultiFormatResource` does not have some corresponding tests.  So it seemed worth changing.

What remains of nevow now (counting changes in this PR) are (1) things in `webish`, (2) things in `web.common`, (3) `inevow.IRequest` in some tests (including this one), (4) `nevow.url` in three places, which looks like should go when we update `webish`.

Now might be a good time to change this?

Reply:
Here, I think, is the risk of this change.

Previously, this function did something like "render a page like the Nevow appserver will render it".  Now it does something more like "render a page like a Twisted Web site will render it (with some Nevow-isms mixed in - the `fields` attribute is only available with Nevow requests, right?)".

The way Tahoe-LAFS web code actually works right now is to be rendered by the Nevow appserver.  So this change removes the test coverage that exercises the web code in the way it is *really* used.  This doesn't invalidate all of the tests for the web code that use this function but it does invalidate anything that's about an interaction between our code and Nevow.

The way Tahoe-LAFS web code will work in the future, we hope, is to be rendered by a Twisted Web site - with no Nevow-isms.  So the new version of this functional also doesn't show us the web code will work as we want after we complete the Nevow transition.

If it doesn't show us the code works now and it doesn't show us the code will work in the future, it doesn't seem like it's very useful.

I think that "does it work now" is the best thing for the test suite to demonstrate so I think the best option is to leave this as it was: exercising Nevow appserver interactions.  When we think we have no more Nevow appserver dependencies and are completely compatible with Twisted Web site then we can make a change like this one but without any of the Nevow-isms.